### PR TITLE
feat: create Exception from tracing event

### PR DESCRIPTION
**Pull request by seanpianka**
This PR relates to my message in #180 

This change will build an exception differently than event. It will look
for two hardcoded keys in the "extras" of an event:
1. "error_kind" - this is optional and defaults to "Unknown"
2. "error" - this is optional and defaults to the event's message

I am open to ideas for improving this! This PR simply gets exception 
handling off the ground, but makes arbitrary assumptions about how the
tracing event will communicate the error information.

### CODE REVIEW

File: sentry-tracing/src/converters.rs [hunk 1]
```diff
@@ -141,9 +142,66 @@ pub fn exception_from_event<S>(event: &tracing_core::Event, ctx: Context<S>) ->
...
+    let exception = Exception {
+        ty,
+        value,
+        module,
+        ..Default::default()
+    };
+    let culprit = match (event.metadata().file(), event.metadata().line()) {
+        (Some(f), Some(l)) => Some(format!("{}L{}", f, l)),
+        _ => None,
+    };
```
Comment (flub) [inline sentry-tracing/src/converters.rs:174]: I've been trying to find out what this field is for and so far the SDK folks seem to be telling me that this is an internal field that's set during processing and also that it is deprecated.  So we shouldn't set it here.
(see also https://github.com/getsentry/relay/pull/1082)

**Review by flub — COMMENTED** [review]
I like the general approach to this, thanks for starting to work on it.  Some high level questions I have though:

* I would kind of expect to be able to pass an `std::error::Error` type in the `error` field of an trancing span/event and for the sentry integration to extract the type and value and backtrace just like it would do for directly capturing such error types.  Can this be done?
* I think the crate's doc comment should clearly document that the integration treats those tracing fields specially and how to use them to create sentry errors.
* I'm not entirely sure if I'm reading this code correctly, but does this try to create an exception for any tracing event?  Should it only do this for error-level events maybe?

File: sentry-tracing/src/converters.rs (hunk 1 above)
Comment (seanpianka) [reply sentry-tracing/src/converters.rs]: Removed!

I tried to shoehorn the filename and line number of the span into the exception somehow, but perhaps this information will be included elsewhere? I am unsure about this, since there's no stacktrace available.

**Comment by seanpianka** [comment]
> * I would kind of expect to be able to pass an `std::error::Error` type in the `error` field of an trancing span/event and for the sentry integration to extract the type and value and backtrace just like it would do for directly capturing such error types.  Can this be done?

Yes, this should be possible. The [values](https://docs.rs/tracing/0.1.26/tracing/trait.Value.html) for a field in a tracing span/event are able to record a `dyn Error + 'static`, so retrieving that from the event _should_ be possible. However, when looking over the documentation for [`#[tracing::instrument(err)]`](https://tracing.rs/tracing/attr.instrument.html), it seems to me that it is expecting the error to be `Display` so that it can be printed in the "error" field of any raised `ERROR` events.

> * I think the crate's doc comment should clearly document that the integration treats those tracing fields specially and how to use them to create sentry errors.

Agreed. My choice of field names is arbitrary, and I'd like to align with whichever fields `#[tracing::instrument(err)]` uses for errors.

> * I'm not entirely sure if I'm reading this code correctly, but does this try to create an exception for any tracing event?  Should it only do this for error-level events maybe?

Perhaps I'm misunderstanding the [`on_event`](https://github.com/getsentry/sentry-rust/blob/master/sentry-tracing/src/layer.rs#L278) method in the layer, but I don't think it's creating an exception for any tracing event, only those that are ERROR level or above.

**Comment by seanpianka** [comment]
So, I've re-worked the `record_error` impl using @leops' fork as [an example](https://github.com/levels3d/sentry-tracing/blob/master/src/converters.rs#L135-L142). Now, issues reported in Sentry have their own exceptions with relevant type, title/message, and a (semi-relevant) backtrace.

The "ty" of the exception is still not great. For an error event, the type is `Event path/to/file.rs:LineNumber`. Better than "Unknown", but still not great.

Additionally, it seems like some "extras" are being lost in the issue being sent to Sentry. In my testing with a function annotated with `#[tracing::instrument(err, fields(data = 1))]`, I am unable to find `data = 1` anywhere in the reported Sentry issue... so some more debugging is required.

File: sentry-tracing/src/converters.rs [hunk 2]
```diff
@@ -22,93 +24,99 @@ pub fn convert_tracing_level(level: &tracing_core::Level) -> Level {
...
+
     fn record_bool(&mut self, field: &Field, value: bool) {
         self.record(field, value);
     }
+
     fn record_str(&mut self, field: &Field, value: &str) {
         self.record(field, value);
     }
+
+    fn record_error(&mut self, _field: &Field, value: &(dyn Error + 'static)) {
```
Comment (flub) [inline sentry-tracing/src/converters.rs:87]: ooh, very nice that tracing provides this!

File: sentry-tracing/src/converters.rs [hunk 3]
```diff
@@ -22,93 +24,99 @@ pub fn convert_tracing_level(level: &tracing_core::Level) -> Level {
...
+        let event = event_from_error(value);
+        for exception in event.exception {
```
Comment (flub) [inline sentry-tracing/src/converters.rs:89]: We should probably provide a better API for this, but this will do for now.

File: sentry-tracing/src/converters.rs [hunk 4]
```diff
@@ -22,93 +24,99 @@ pub fn convert_tracing_level(level: &tracing_core::Level) -> Level {
...
+            self.exceptions.push(exception);
```
Comment (flub) [inline sentry-tracing/src/converters.rs:90]: I think you can use `self.exceptions.append(event.exception)` instead of using a loop.  that's implemented a little more efficiently

File: sentry-tracing/src/converters.rs [hunk 5]
```diff
@@ -145,5 +173,37 @@ where
...
+    let (message, mut visitor) = extract_event_data(event);
+
+    let exception = if visitor.exceptions.is_empty() {
+        let ty = event.metadata().name().into();
+        let value = visitor
+            .json_values
+            .remove(DEFAULT_ERROR_VALUE_KEY)
+            .map(|v| v.as_str().map(|s| s.to_owned()))
+            .flatten()
+            .or_else(|| message.clone());
```
Comment (flub) [inline sentry-tracing/src/converters.rs:187]: Is this whole thing still needed now that you discovered `record_error`?

**Review by flub — COMMENTED** [review]
Apologies for the delay in getting back!

File: sentry-tracing/src/converters.rs (hunk 1 above)
Comment (flub) [reply sentry-tracing/src/converters.rs]: Yes, it would be nice to have this somewhere.  But indeed without a stacktrace that seems tricky.  I'm not very familiar with the event data format, https://develop.sentry.dev/sdk/event-payloads/#other-interfaces might have info on where else could be suitable.

File: sentry-tracing/src/converters.rs (hunk 4 above)
Comment (seanpianka) [reply sentry-tracing/src/converters.rs]: Fixed, thanks!

File: sentry-tracing/src/converters.rs (hunk 5 above)
Comment (seanpianka) [reply sentry-tracing/src/converters.rs:185]: I left this branch as a fallback in the case that the event does not have any exceptions. This would at least create a single exception using what data is available from either the `error` or `message` field (in that order).

File: sentry-tracing/src/converters.rs (hunk 3 above)
Comment (seanpianka) [reply sentry-tracing/src/converters.rs]: You could create a separate issue for it, if you have a description in mind. "Sentry event exceptions should not be exposed through a public field?"

File: sentry-tracing/Cargo.toml [hunk 6]
```diff
@@ -12,7 +12,8 @@ Sentry integration for tracing and tracing-subscriber crates.
 edition = "2018"
 
 [dependencies]
-sentry-core = { version = "0.23.0", path = "../sentry-core" }
+sentry-core = { version = "0.23.0", path = "../sentry-core", features = ["client"] }
+sentry-backtrace = { version = "0.23.0", path = "../sentry-backtrace" }
```
Comment (Swatinem) [inline sentry-tracing/Cargo.toml:16]: why was this change necessary? I don’t see these imports being used? Please revert this in that case.

**Review by Swatinem — COMMENTED** [review]
lgtm, but it would be nice to have a test for this.

**Comment by Swatinem** [comment]
I think a simple test where you supply some kind of `Error` type, and have that be captured correctly by sentry would be nice.

**Comment by Swatinem** [comment]
@seanpianka 👋🏻 what is still needed to push this over the finish line?

I might be a bit tempted to just merge this as is, as we are currently adding manual performance monitoring support and will then probably overhaul the tracing integration completely.

**Comment by Swatinem** [comment]
I went ahead and merged the code with the latest upstream changes in #412
